syntax = "proto3";
package csce438;
import "google/protobuf/timestamp.proto";

// ------------------------------------------------------------
// The coordinator service definition
// These are the different functionalities exposed by service
// ------------------------------------------------------------

//Init and Heartbeat potentially redundant
service CoordService{
  rpc Heartbeat (ServerInfo) returns (Confirmation) {}
  rpc GetServer (ID) returns (ServerInfo) {}
  // ZooKeeper API here
  rpc GetAllFollowerServers (ID) returns (ServerList) {}
  rpc GetFollowerServer (ID) returns (ServerInfo) {}
  rpc Init (ServerInfo) returns (Confirmation) {}
}

//server info message definition
message ServerInfo{
  int32 serverID = 1;
  string hostname = 2;
  string port = 3;
  string type = 4;
}

//confirmation message definition
message Confirmation{
  bool status = 1;
  string designation = 2;
  string worker_hostname = 3;
  string worker_port = 4;
}

//id message definition
message ID{
  int32 id = 1;
}

message ServerList{
  repeated int32 serverID = 1;
  repeated string hostname = 2;
  repeated string port = 3;
  repeated string type = 4;
}


service SynchService{
  rpc GetAllUsers (Confirmation) returns (AllUsers) {}
  rpc GetTLFL (ID) returns (TLFL) {}
  rpc ResynchServer (ServerInfo) returns (Confirmation) {}
}

message TLFL{
  bool status = 1;
  repeated string TL = 2;
  repeated string FL = 3;
}

message AllUsers{
  repeated string users = 1;
}
